### ISSUE TITLE
wagtail.core.models.Site is rejecting certain Page objects.

### ISSUE THREAD

**Issue by alvindera97** [wagtail/wagtail#6461]
Found a bug? Please fill out the sections below. 👍

### Issue Summary

**_A summary of the issue._**
Internal Server Error is displayed while running tests even though the Site object saves. 

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. Edit **models.py** as follows...

• **home / models.py**
```
class HomePage(Page):
    content = StreamField([
        ('our_services_statement', StatementBlock()),
        ('the_technology_we_utilize', TechStatementBlock()),
        ('meet_codedan_developers', DeveloperBlock()),
        ('meet_codedan_analysts', AnalystsBlock()),
    ], null=True, blank=True)
    objects = models.PageManager()

    content_panels = Page.content_panels + [
        StreamFieldPanel('content'),
    ]
    template = 'home.html'
```

3. Edit **home/tests.py** as follows:

• **home / tests.py**

```
class HomePageTestCase(StaticLiveServerTestCase, TestCase):
    @classmethod
    def setUpClass(cls):
        if Page.objects.first():
            cls.root = Page.objects.create(title="Root", path='001', depth=1)
        else:
            cls.root = Page.objects.get(title="Root").specific

        # Index page is compulsory
        cls.codedan_index_page = HomePage(
            title="Home Page",
            content = [
                    ('our_services_statement', {'statement': '<b>Default</b> statement'}),
                    ('the_technology_we_utilize', {'statement': '<b>Default</b> technology'}),
                    ('meet_codedan_developers', {'statement': '<b>Default</b> developers'}),
                    ('meet_codedan_analysts', {'statement': '<b>Default</b> analysts'}),
            ]
        )
        cls.root.add_child(instance=cls.codedan_index_page)
        cls.saved_revision_codedan_index_page = cls.codedan_index_page.save_revision()
        cls.saved_revision_codedan_index_page.publish()
        cls.codedan_index_page.save()

        # Create site object if there isn't any

        if Site.objects.first():
            cls.site = Site.objects.first()
            cls.site.root_page = cls.codedan_index_page
            cls.site.save()

        else:
            cls.site = Site.objects.create(
                site_name="localhosted",
                is_default_site=True,
                root_page=cls.codedan_index_page
            )

        cls.codedan_index_page.content = [
            ('our_services_statement', {'statement': '<b>Default</b> statement'}),
            ('the_technology_we_utilize', {'statement': '<b>Default</b> technology'}),
            ('meet_codedan_developers', {'statement': '<b>Default</b> developers'}),
            ('meet_codedan_analysts', {'statement': '<b>Default</b> analysts'}),
        ]

        cls.codedan_index_page.save()
        super(HomePageTestCase, cls).setUpClass()
        cls.browser = selenium.webdriver.Chrome()

    def test_tech_we_utilize_gets_displayed(self):
        self.browser.get(self.live_server_url)
        technology_we_utilize = self.browser.find_element_by_id('tech-we-use-chunk')
        self.assertTrue(technology_we_utilize.text.is_displayed(), True)

```

Any other relevant information. For example, why do you consider this a bug and what did you expect to happen instead?
• I consider this a bug because if you set the `root_page` of the `Site` object to the root page (`Page.objects.get(id=1)`), there is no server error displayed. However, the root page isn't the desired Index Page. This means that calls to the index page `self.get_live_server_url` would load a web page that just displays the text **"Root"**  or the familiar **"Welcome to your wagtail site"** page. This is the code to explain what I am describing:

```
# in this case, there is no Internal Server Error Displayed
class HomePageTestCase(StaticLiveServerTestCase, TestCase):
        # ...
        if Site.objects.all():
            cls.site = Site.objects.first()
            cls.site.root_page = cls.root
            cls.site.save()

        else:
            cls.site = Site.objects.create(
                site_name="localhosted",
                is_default_site=True,
                root_page=cls.root
            )
        # ...

```

• I also consider this a bug because, if you don't add the stream field information, and set the desired index page as the `root_page` of the `Site` object, it loads with no problem. However, the `HomePage` object does not have any StreamField data saved (which defeats the aim of having a test suite in the first place). Below is the code describing what I have just described.

```
# this is another situation where there is no Internal Server Error displayed
class HomePageTestCase(StaticLiveServerTestCase, TestCase):
        # ...

        # Index page is compulsory
        cls.codedan_index_page = HomePage(  # notice that the StreamField data has been removed
            title="Home Page",
        )
        cls.root.add_child(instance=cls.codedan_index_page)
        cls.saved_revision_codedan_index_page = cls.codedan_index_page.save_revision()
        cls.saved_revision_codedan_index_page.publish()
        cls.codedan_index_page.save()

        # Create site object if there isn't any

        if Site.objects.all():
            cls.site = Site.objects.first()
            cls.site.root_page = cls.codedan_index_page
            cls.site.save()

        else:
            cls.site = Site.objects.create(
                site_name="localhosted",
                is_default_site=True,
                root_page=cls.codedan_index_page
            )
        # ...

```

SO FROM ALL INDICATIONS, THE SITE OBJECT IS FINE WITH HAVING `root_page` SET TO A ROOT PAGE OBJECT (e.g. **Page.objects.get(id=1)**)  OR HAVING root_page SET TO ANOTHER PAGE OBJECT (IN THIS CASE THE INDEX OBJECT) _AS LONG AS THAT PAGE OBJECT DOES NOT HAVE STREAMFIELD DATA SET UP_

_Please also note that in all these situations, the_ `Site` _object still saves but the test case returns a browser window saying **Internal Server Error**_

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: (no)

### Technical details

* Python version: **Python 3.6.9**
* Django version: **3.1.2**
* Wagtail version: **2.10.2**
* Browser version: https://www.whatsmybrowser.org/b/Y8H8M

**Comment by tomdyson** [comment]
Hi @alvindera97, thanks for your detailed report. 

Does your `HomePage` model work fine in the admin UI: can you create / preview / publish a page? Does `home.html` exist?

You haven't included the definition of your various Blocks here. Is `{'statement': '<b>Default</b> developers'}` a workable value for all of them?

Have you [configured logging](https://docs.djangoproject.com/en/3.1/topics/logging/) so that you can get some more helpful error output from your tests?

**Comment by gasman** [comment]
Hi @alvindera97,
The steps you have provided are not sufficient to replicate the issue, as they do not include the template or StreamField block definitions. With some basic adjustments (such as adding `<div id="tech-we-use-chunk">` to the home.html template) I was able to get the test to pass, which suggests that the problem is at least partly in the code we haven't seen here.

As I mentioned at https://github.com/wagtail/wagtail/issues/6455#issuecomment-707697173, the failure of this testing setup to provide a meaningful error message is a Selenium / Django configuration issue, not a Wagtail bug. If you were to rewrite the test using [the Django test client](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#default-test-client) to fetch the appropriate page and check for the ID in the HTML, I'm confident that it would give you a more useful error trace. That doesn't mean that Selenium is the wrong approach - just that there is more setup to be done that's out of scope for us to assist with. (Yes, it would indeed be nice if this all worked out of the box, but as I'm sure you understand, we have enough work to do in maintaining Wagtail itself - how other parts of the tech stack interoperate is really out of our hands.)

**Comment by alvindera97** [comment]
> Hi @alvindera97, thanks for your detailed report.
> 
> Does your `HomePage` model work fine in the admin UI: can you create / preview / publish a page? Does `home.html` exist?
> 
> You haven't included the definition of your various Blocks here. Is `{'statement': '<b>Default</b> developers'}` a workable value for all of them?
> 
> Have you [configured logging](https://docs.djangoproject.com/en/3.1/topics/logging/) so that you can get some more helpful error output from your tests?

Yeah, my `HomePage` model works fine in the admin UI. I wasn't able to notice anything amiss with the `setUpClass` until I tried running the test. I think this all boils down to the inability to get a proper glimpse of what's wrong during the test. The "**Internal Server Error**" page is not only annoying, It is extremely useless at anything other than claiming that there was a server error. 

Thank you for the hard work and support that you have put into Wagtail and into Open Source Development. I am really proud to be working with Wagtail and Wagtail Developers (I also think you guys have a cooler "It worked" page counterpart compared to Django's" ... You've all been [unsurprisingly] right with your recommendations and it is now much clearer to me that this isn't a Wagtail specific problem. Unfortunately, I am at a dilemma on who exactly I should lay my complaint to — 
 "Selenium" or "Django" — in reference to the "**Internal Server Error**" page. If you have any suggestions, I would love to hear it. _It would make the world a better place if a trace back or an error page was displayed_.

To the solution of the problem, it was basically a syntax error on the template. I couldn't believe it but it turned out that everything worked apart from the section of my template that was dependent on the StreamField content. Another thing that made this issue look like a bug was that when I used `@override_settings(DEBUG=True)` on the test method, there was no TemplateSyntaxError page displayed. Still the same old "**Internal Server Error**". 

Thanks again. If there's any idea as to who I can report the "Internal System Error" page to, or if you think it's worth reporting, I would really love to hear it. It's been a pleasure working with the community and Wagtail Developers.
